pkg/controller/elasticsearch/driver/upgrade_predicates_test.go

```diff
@@ -367,6 +378,140 @@ func TestUpgradePodsDeletion_Delete(t *testing.T) {
 			wantErr:                      false,
 			wantShardsAllocationDisabled: true,
 		},
+		{
+			name: "stop deletion if yellow in a 2 node cluster and shard has replica unassigned, one node already upgraded",
+			fields: fields{
+				esVersion: "7.5.0",
+				upgradeTestPods: newUpgradeTestPods(
+					newTestPod("masters-0").isMaster(true).isData(true).isHealthy(true).needsUpgrade(true).isInCluster(true).withVersion("7.4.0"),
+					newTestPod("masters-1").isMaster(true).isData(true).isHealthy(true).needsUpgrade(true).isInCluster(true).withVersion("7.5.0"),
+				),
+				maxUnavailable: 1,
+				numberOfPods:   2,
+				health:         client.Health{Status: esv1.ElasticsearchYellowHealth, NumberOfNodes: 2},
+				shardLister: migration.NewFakeShardLister(client.Shards{
+					// One shard is not assigned on masters-0 because its version is not the expected one
```

Comment:
This is interesting, but I am not sure I understand the scenario. You mean the version of the index requires 7.4.0, but the only other node is already 7.5.0? This would be a deadlocked upgrade then? I would expect Elasticsearch to have primary on `masters-1` and replica on `masters-0` or if it is the other way round promote the replica to primary and unassigned the replica on `masters-0` because it is now potentially no longer compatible with the upgrade version of the index. But I am not 100% sure myself how this works internally in Elasticsearch.

Reply:
Hey @pebrc , yes, the test and the name are not really clear, let me elaborate more.
I had a few issues (ESS or on-prem) with the new "Data tiers" feature, so I was thinking about how this could have an impact on ECK.
I have seen that some users (small clusters) have a few hot nodes along with one node in warm where they keep old data with no replicas.
My use case is the following: there is a cluster with 2 hot nodes and 1 warm node, when upgrading the warm node works but we lose one of the hot nodes, then the shard could not be assigned to the warm node, so we move to yellow.
So, the cluster could not finish the upgrade, as we only would have one shard/copy of the `index_a`.

what do you think about this? I'm not sure if my explanation makes sense.

I updated the name of the test and the pods name too, to better reflect the use case

Thanks

Reply:
I see what you mean now. But I think your test is not testing for that. Because you have specified your test cluster as `masters-hot-0` and `masters-warm-1` the test code assumes one of the warm nodes is missing (`masters-warm-0`) not one of the hot nodes. That means we have one unavailable node (`masters-warm-0`) and the `do_not_restart_healthy_node_if_MaxUnavailable_reached` predicate applies. Also I would assume that `needsUpgrade` should be `false` on the node already upgraded.

I would suggest to simplify the test and remove the notion of data tiers as the predicate system is not aware of them anyway. Rather I would rewrite the test to cover the scenario where we have a multi-node cluster where one Pod is temporarily unavailable (because of host failure or some other reason) and where we want to make sure that the single node exception you added does not apply in this case.

Reply:
yes agree, I updated the code reviewed the comments and settings, and the test is now a two-node cluster one pod is unavailable/unhealthy, we are in the situation only one node on the cluster so the shard is assigned but not the replica, so that could match the single node exception (but not because we are using StatefulSetList) so this is not a really single node cluster so we can not delete the pod. 

Thanks for your time and patience Peter